logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/Handler.java

```diff
@@ -0,0 +1,71 @@
+package org.zalando.logbook.jaxws;
+
+import lombok.AllArgsConstructor;
+import org.zalando.logbook.HttpRequest;
+import org.zalando.logbook.HttpResponse;
+import org.zalando.logbook.Logbook;
+import org.zalando.logbook.Logbook.ResponseProcessingStage;
+import org.zalando.logbook.Origin;
+
+import javax.annotation.Nonnull;
+import javax.xml.ws.handler.soap.SOAPMessageContext;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+import static javax.xml.ws.handler.MessageContext.MESSAGE_OUTBOUND_PROPERTY;
+
+@AllArgsConstructor
+final class Handler {
+
+    private static final String ATTRIBUTE = ResponseProcessingStage.class.getName();
+
+    private final Logbook logbook;
+    private final SOAPMessageContext context;
+
+    boolean isIncoming() {
+        return !isOutgoing();
+    }
+
+    boolean isOutgoing() {
+        return (boolean) context.get(MESSAGE_OUTBOUND_PROPERTY);
+    }
+
+    void handleRequest(final Origin origin) {
+        final Request request = new Request(new HttpMessageContext(context), origin);
+        final ResponseProcessingStage stage = write(request);
+        memorize(stage);
+    }
+
+    void handleResponse(final Origin origin) {
+        final ResponseProcessingStage stage = recall();
+        write(stage, new Response(new HttpMessageContext(context), origin));
+    }
+
+    private void memorize(final ResponseProcessingStage stage) {
+        context.put(ATTRIBUTE, stage);
```

Comment:
I have seen a similar strategy across the library, but this expose the PresponseProcessingStage under context/request.
Using an alternative of ThreadLocal allows to hide any library concept, avoiding get the reference of ResponseProcessingStage in any place of an application with access to javax.xml.ws.handler.soap.SOAPMessageContext/javax.servlet.http.HttpServletRequest.

Reply:
Yes, technically it's being exposed, but those facilities exist exactly for our use case, to store context information between requests and responses. ThreadLocal is a smell, that I'm trying to avoid, because I'd rather delegate the job of keeping track of the context to the framework, then imposing  the requirement to have the same thread process the request and response (which is not ensured).